common/src/main/java/io/pravega/common/util/btree/BTreeIndex.java

```diff
@@ -0,0 +1,825 @@
+ * the root page). Each such modification will result in all these "touched" pages to be rewritten to the external data source.
+ * * Upon writing, the contents of the Index Pages will be updated to reflect the new offsets of their direct child pages.
+ * * Upon writing, a fixed-length "footer" is written at the very end which contains information about where the most recent
+ * root page is located. This eliminates the need for such a pointer to be stored externally or to have a fixed-length root
+ * page, which would be wasteful.
+ *
+ * Read-write consistency and concurrency:
+ * * Due to the append-only nature of the storage media, reads (including iterators) may occur concurrently with writes,
+ * without having to worry about consistency. This is because when a read is initiated, it caches the root page (along
+ * with any other sub-pages), which will always point to locations with smaller offsets (by construction). Any new writes
+ * (concurrent or not) will always operate on higher offsets, thus not interfering with any ongoing reads.
+ * * This implementation does not support concurrent writes. Writes will need to be serialized, either within the same
+ * instance or across multiple instances. Concurrent writes will be guarded by means of "appending-at-offset" within the
+ * external data source (when the BTreeIndex is first used, it gets a snapshot of the external data source which it then
+ * uses for conditionally writing modifications).
+ * * If two or more concurrent writes occur on the same BTreeIndex instance, only one will succeed (and the others may be
+ * retried).
+ * * If two or more concurrent writes occur on different BTreeIndex instances, only one will succeed. The "losing" instances
+ * will become unavailable for writing and will present an outdated view of the data for reading. This mechanism allows
+ * the caller to decide how to properly recover from the situation - the BTreeIndex has insufficient information to make
+ * such a decision.
+ *
+ * Versioning:
+ * * BTreePages have built-in versioning; please refer to the BTreePage class for details. It is possible to mix different
+ * BTreePage versions in the same BTreeIndex structure.
+ * * BTreeIndex has no built-in versioning, as we would not be able to mix different versions of the BTreeIndex in the same
+ * data source - that is, we cannot begin writing at version X, then after a while we switch to version Y in the same file.
+ * * For BTreeIndex versioning (when it will be needed), a suggested approach is to pass in the version via the constructor
+ * which should tell the BTreeIndex how to interpret the data in the external data source. Once a BTreeIndex is written in
+ * one version in a file, it can only be "upgraded" if it is bulk-loaded into a different file (such a feature is not yet
+ * supported). This versioning would have to be maintained externally (i.e., in a Segment Core Attribute or by file naming
+ * conventions).
+ */
+@NotThreadSafe
```

Comment:
It's not thread safe because it has undefined behavior in case of concurrent write access. The members themselves are thread safe, and the index will handle concurrent requests without corrupting the index, but there is no guarantee as to what will happen if two or more modify requests come in at the same time. Even if A starts and returns synchronously before B is invoked, it may just be that B is applied first and A will be failed (or viceversa). That's what I meant by this tag.